File: tests/netDxf.Conformance/MeshBlendCreaseTests.cs
using netDxf;
using netDxf.Blocks;
using netDxf.Entities;
using netDxf.Header;
using netDxf.IO;

namespace NetDxf.Conformance;

internal static partial class Program
{
    private static void RegisterMeshBlendCreaseTests()
    {
        Run("mesh/blend/default", () => Check(!ProfileMesh(0).BlendCrease,"Default mesh blend flag is not false."));
        foreach (DxfVersion version in SupportedVersions.Where(v => v >= DxfVersion.AutoCad2010))
            foreach (bool binary in new[] { false, true })
            {
                DxfVersion v = version; bool b = binary;
                foreach (bool enabled in new[] { false, true })
                {
                    bool e = enabled;
                    Run($"mesh/blend/api-clone-explode/{v}/{b}/{e}", () => MeshBlendApi(v,b,e));
                }
                foreach (short? value in new short?[] { null, 0, 1 })
                    foreach (bool late in new[] { false, true })
                    {
                        short? f = value; bool l = late;
                        Run($"mesh/blend/read-write/{v}/{b}/{f}/{l}", () => MeshBlendWire(v, b, f, l));
                    }
                foreach (short value in new short[] { -1, 2 })
                {
                    short f = value;
                    Run($"mesh/blend/invalid/{v}/{b}/{f}", () => MeshBlendInvalid(v, b, f));
                }
            }
    }

    private static MemoryStream MeshBlendFixture(DxfVersion version, bool binary, short? flag, bool late)
    {
        var tags = new List<DxfTag>
        {
            new(0,"SECTION"), new(2,"HEADER"), new(9,"$ACADVER"), new(1,HeaderVersion(version)), new(0,"ENDSEC"),
            new(0,"SECTION"), new(2,"ENTITIES"), new(0,"MESH"), new(5,"200"), new(100,"AcDbEntity"), new(8,"0"),
            new(100,"AcDbSubDMesh"), new(71,(short)2)
        };
        void Flag()
        {
            if (flag.HasValue) tags.Add(new DxfTag(72,flag.Value));
            if (!binary) tags.Add(new DxfTag(999,"interleaved comment"));
        }
        if (!late) Flag();
        tags.AddRange(new DxfTag[]
        {
            new(91,2), new(92,3),
            new(10,0.0),new(20,0.0),new(30,0.0),new(10,1.0),new(20,0.0),new(30,0.0),new(10,0.0),new(20,1.0),new(30,0.0),
            new(93,4),new(90,3),new(90,0),new(90,1),new(90,2),
            new(94,1),new(90,0),new(90,1),new(95,1),new(140,1.5),new(90,0)
        });
        if (late) Flag();
        tags.AddRange(new DxfTag[] { new(1001,"MESH_BLEND"),new(1000,"after mesh"),new(0,"ENDSEC"),new(0,"EOF") });
        return new MemoryStream(RawFixtureBytes(tags,binary));
    }

    private static void MeshBlendWire(DxfVersion version, bool binary, short? flag, bool late)
    {
        using var input = MeshBlendFixture(version,binary,flag,late);
        var document = DxfDocument.Load(input) ?? throw new InvalidOperationException("MESH blend fixture failed to load.");
        Mesh original = document.Entities.Meshes.Single();
        Equal((byte)2,original.SubdivisionLevel,"Blend flag affected subdivision level");
        Equal(3,original.Vertexes.Count,"Blend flag affected vertices");
        Equal(1.5,original.Edges.Single().Crease,"Blend flag affected crease value");
        Equal("after mesh",(string)original.XData["MESH_BLEND"].XDataRecord.Single().Value,"Blend flag disrupted XData");
        var copy = (Mesh)original.Clone(); document.Entities.Add(copy);
        using var output = new MemoryStream(); Check(document.Save(output,!binary),"MESH blend export failed.");
        output.Position = 0; var raw = DxfRawDocument.Load(output);
        var meshes = raw.Sections.SelectMany(s=>s.Records).Where(r=>r.Name=="MESH").ToArray();
        Equal(2,meshes.Length,"Cloned MESH count");
        foreach (var mesh in meshes)
            Equal(flag ?? (short)0,(short)mesh.Tags.Single(t=>t.Code==72).Value,"MESH blend flag was reset");
        Check(input.CanRead && output.CanWrite,"MESH blend IO closed caller streams.");
        if (flag==1 && !late)
            File.WriteAllBytes(Path.Combine(ArtifactDirectory,$"mesh-blend-{version}-{!binary}.dxf"),output.ToArray());
    }

    private static void MeshBlendApi(DxfVersion version, bool binary, bool enabled)
    {
        Mesh mesh = ProfileMesh(2); mesh.BlendCrease = enabled;
        var block = new Block("BlendMesh"); block.Entities.Add(mesh);
        var insert = new Insert(block);
        var copy = (Insert)insert.Clone();
        Mesh child = copy.Block.Entities.OfType<Mesh>().Single();
        Equal(enabled,child.BlendCrease,"Block clone lost blend flag");
        child.BlendCrease = !enabled;
        Equal(enabled,mesh.BlendCrease,"Clone edit changed the source blend flag");
        Mesh exploded = insert.Explode().OfType<Mesh>().Single();
        Equal(enabled,exploded.BlendCrease,"Explosion lost blend flag");
        var document = new DxfDocument(version); document.Entities.Add(insert);
        using var output = new MemoryStream(); Check(document.Save(output,binary),"Authored blend flag failed to save.");
        output.Position = 0;
        var loaded = DxfDocument.Load(output) ?? throw new InvalidOperationException("Authored blend flag failed to load.");
        Equal(enabled,OnlyProfileMesh(loaded).BlendCrease,"Authored blend flag changed on reload");
    }

    private static void MeshBlendInvalid(DxfVersion version, bool binary, short flag)
    {
        using var input = MeshBlendFixture(version,binary,flag,false);
#if DEBUG
        try { DxfDocument.Load(input); throw new InvalidOperationException("Invalid blend flag was accepted."); }
        catch (InvalidDataException exception)
        {
            Check(exception.Message.Contains("72",StringComparison.Ordinal),"Blend flag diagnostic lost the group code.");
        }
#else
        Check(DxfDocument.Load(input)==null,"Invalid blend flag was accepted.");
#endif
        Check(input.CanRead,"Invalid MESH input closed caller stream.");
    }
}

File: javascript/tests/netDxf.Conformance/MeshBlendCreaseTests.js
// Port of every original case in pinned MeshBlendCreaseTests.cs, retaining the default identity.
import { DxfDocument, DxfRawDocument, DxfTag, DxfVersion, Mesh, Block, Insert, MemoryStream } from '../../index.js';
import { InvalidDataException } from '../../runtime/Errors.js';
import { GetTypedIOConfiguration } from '../../runtime/TypedDocumentIO.js';
import { Run, Check, Equal, SupportedVersions, VersionName, HeaderVersion, BooleanName } from './TestHarness.js';
import { RawFixtureBytes } from './RawDocumentTests.js';
import { ProfileMesh, OnlyProfileMesh, MeshSingle, MeshRecords, MeshWriteArtifact } from './MeshVersionTests.js';
export function RegisterMeshBlendCreaseTests(){
  Run('mesh/blend/default',()=>Check(!ProfileMesh(0).BlendCrease,'Default mesh blend flag is not false.'));
  for(const v of SupportedVersions.filter(v=>v>=DxfVersion.AutoCad2010))for(const b of [false,true]){
    const id=`${VersionName(v)}/${BooleanName(b)}`;
    for(const e of [false,true])Run(`mesh/blend/api-clone-explode/${id}/${BooleanName(e)}`,()=>MeshBlendApi(v,b,e));
    for(const f of [null,0,1])for(const late of [false,true])Run(`mesh/blend/read-write/${id}/${f??''}/${BooleanName(late)}`,()=>MeshBlendWire(v,b,f,late));
    for(const f of [-1,2])Run(`mesh/blend/invalid/${id}/${f}`,()=>MeshBlendInvalid(v,b,f));
  }
}
export function MeshBlendFixture(v,b,flag,late){
  const tags=[[0,'SECTION'],[2,'HEADER'],[9,'$ACADVER'],[1,HeaderVersion(v)],[0,'ENDSEC'],[0,'SECTION'],[2,'ENTITIES'],[0,'MESH'],[5,'200'],[100,'AcDbEntity'],[8,'0'],[100,'AcDbSubDMesh'],[71,2]].map(([c,x])=>new DxfTag(c,x));
  const addFlag=()=>{if(flag!==null)tags.push(new DxfTag(72,flag));if(!b)tags.push(new DxfTag(999,'interleaved comment'));};if(!late)addFlag();
  tags.push(...[[91,2],[92,3],[10,0],[20,0],[30,0],[10,1],[20,0],[30,0],[10,0],[20,1],[30,0],[93,4],[90,3],[90,0],[90,1],[90,2],[94,1],[90,0],[90,1],[95,1],[140,1.5],[90,0]].map(([c,x])=>new DxfTag(c,x)));if(late)addFlag();
  tags.push(...[[1001,'MESH_BLEND'],[1000,'after mesh'],[0,'ENDSEC'],[0,'EOF']].map(([c,x])=>new DxfTag(c,x)));return new MemoryStream(RawFixtureBytes(tags,b));
}
export function MeshBlendWire(v,b,flag,late){const input=MeshBlendFixture(v,b,flag,late),output=new MemoryStream();try{const doc=DxfDocument.Load(input);Check(doc!==null,'MESH blend fixture failed to load.');const original=MeshSingle(doc.Entities.Meshes);Equal(2,original.SubdivisionLevel,'Blend flag affected subdivision level');Equal(3,original.Vertexes.Count,'Blend flag affected vertices');Equal(1.5,MeshSingle(original.Edges).Crease,'Blend flag affected crease value');Equal('after mesh',MeshSingle(original.XData.get_Item('MESH_BLEND').XDataRecord).Value,'Blend flag disrupted XData');doc.Entities.Add(original.Clone());Check(doc.Save(output,!b),'MESH blend export failed.');output.Position=0;const meshes=MeshRecords(DxfRawDocument.Load(output)).filter(r=>r.Name==='MESH');Equal(2,meshes.length,'Cloned MESH count');for(const mesh of meshes)Equal(flag??0,MeshSingle(Array.from(mesh.Tags).filter(t=>t.Code===72)).Value,'MESH blend flag was reset');Check(input.CanRead&&output.CanWrite,'MESH blend IO closed caller streams.');if(flag===1&&!late)MeshWriteArtifact(`mesh-blend-${VersionName(v)}-${BooleanName(!b)}.dxf`,output.ToArray());}finally{input.Dispose();output.Dispose();}}
export function MeshBlendApi(v,b,enabled){const mesh=ProfileMesh(2);mesh.BlendCrease=enabled;const block=new Block('BlendMesh');block.Entities.Add(mesh);const insert=new Insert(block),copy=insert.Clone(),child=MeshSingle(Array.from(copy.Block.Entities).filter(e=>e instanceof Mesh));Equal(enabled,child.BlendCrease,'Block clone lost blend flag');child.BlendCrease=!enabled;Equal(enabled,mesh.BlendCrease,'Clone edit changed the source blend flag');Equal(enabled,MeshSingle(Array.from(insert.Explode()).filter(e=>e instanceof Mesh)).BlendCrease,'Explosion lost blend flag');const doc=new DxfDocument(v),output=new MemoryStream();doc.Entities.Add(insert);try{Check(doc.Save(output,b),'Authored blend flag failed to save.');output.Position=0;const loaded=DxfDocument.Load(output);Check(loaded!==null,'Authored blend flag failed to load.');Equal(enabled,OnlyProfileMesh(loaded).BlendCrease,'Authored blend flag changed on reload');}finally{output.Dispose();}}
export function MeshBlendInvalid(v,b,flag){const input=MeshBlendFixture(v,b,flag,false);try{if(GetTypedIOConfiguration()==='Debug'){let caught=false;try{DxfDocument.Load(input);}catch(e){Check(e instanceof InvalidDataException,'Expected InvalidDataException');Check(e.message.includes('72'),'Blend flag diagnostic lost the group code.');caught=true;}Check(caught,'Invalid blend flag was accepted.');}else Check(DxfDocument.Load(input)===null,'Invalid blend flag was accepted.');Check(input.CanRead,'Invalid MESH input closed caller stream.');}finally{input.Dispose();}}
